### PR TITLE
Make upgrade orchestrate more specific to upgrade and add downgrade orchestrate

### LINKED ISSUE #1343: Create downgrade orchestrate
**Component**:

'salt', 'upgrade', 'kubernetes'

**Why this is needed**:

Currently we only have an orchestrate for upgrade but we also need one for downgrade.
The downgrade process is the same as the upgrade one, just in reverse order

**What should be done**:

Create an orchestrate for downgrade

**Implementation proposal** (strongly recommended):

Same as upgrade runner in reverse order

### PR DESCRIPTION
**Pull request by TeddyAndrieux**
**Component**:

'salt', 'kubernetes', 'upgrade'

**Context**: 

We want to be able to do a downgrade

**Summary**:

Refactor upgrade orchestrate to be more specific to ugprade and create a downgrade orchestrate

**Acceptance criteria**: 

Have a working downgrade.

To make new iso available
https://github.com/scality/metalk8s/blob/development/2.0/docs/operation/preparation.rst

Process to downgrade here:
https://github.com/scality/metalk8s/issues/1362

WARN: Currently pods not get reschedulled due to #1408 

---

Fixes: #1343

### CODE REVIEW

**Comment by TeddyAndrieux** [comment]
@bert-e reset

**Comment by TeddyAndrieux** [comment]
@bert-e reset

**Comment by TeddyAndrieux** [comment]
@bert-e reset

**Review by slaperche-scality — COMMENTED** [review]
Commit title is too long: `salt: split upgrade and downgrade orchestrate` should fit.

**Comment by TeddyAndrieux** [comment]
@bert-e reset

File: salt/metalk8s/orchestrate/downgrade/precheck.sls [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+{%- set dest = (dest_version|string).split('.') %}
+{%- for node, values in pillar.metalk8s.nodes.items() %}
+
+  {#- As we don't know how many minor version we do before a new major,
+      never block downgrade between major version #}
+  {%- set current = (values['version']|string).split('.') %}
+  {%- if dest[0]|int == current[0]|int
+     and current[1]|int - dest[1]|int > 1 %}
+
+Unable to downgrade from more than 1 version, Node {{ node }} from {{ values['version'] }} to {{ dest_version }}:
```
Comment (jbertran) [inline salt/metalk8s/orchestrate/downgrade/precheck.sls]: `from more than 1 minor version`

File: salt/metalk8s/orchestrate/downgrade/precheck.sls [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+Node {{ node }} ignored, already in {{ values['version'] }} older than {{ dest_version }}:
+  test.succeed_without_changes
+
+  {%- elif dest_version != values['version']|string %}
+
+Node {{ node }} will be downgraded from {{ values['version'] }} to {{ dest_version }}:
+  test.succeed_without_changes
+
+  {%- else %}
```
Comment (jbertran) [inline salt/metalk8s/orchestrate/downgrade/precheck.sls]: For accuracy's sake we could use `{%- elif salt.pkg.version_cmp(dest_version, values['version']) == 1` here and a relevant failing state name in the `{%- else %}`, where we should theoretically never end up.

Also, I'm not sure I have all the pieces on expected version values but what if we have some more trailing stuff? Do we treat `2.1.1-pw` like we do `2.1.1`? (see our current 2.0.0**-dev**)

File: salt/metalk8s/orchestrate/etcd.sls [hunk 3]
```diff
@@ -33,7 +33,7 @@ Check etcd cluster health for {{ node }}:
       - minion_id: {{ node }}
     - attemps: 5
```
Comment (jbertran) [inline salt/metalk8s/orchestrate/etcd.sls:34]: Unrelated to your PR: we should correct our `module.run` to use the keyword `attempts` and not `attemps`.

File: salt/metalk8s/orchestrate/downgrade/precheck.sls (hunk 2 above)
Comment (jbertran) [reply salt/metalk8s/orchestrate/downgrade/precheck.sls]: Talking with @gdemonet , treating versions with labels the same as versions without is fine for me. Maybe we can strip labels when we compare versions, in order to actually check `0`, `-1` and `1` explicitly?

File: salt/metalk8s/orchestrate/etcd.sls (hunk 3 above)
Comment (TeddyAndrieux) [reply salt/metalk8s/orchestrate/etcd.sls:34]: Oh indeed, feel free to create a ticket :stuck_out_tongue:

File: salt/metalk8s/orchestrate/downgrade/precheck.sls (hunk 1 above)
Comment (TeddyAndrieux) [reply salt/metalk8s/orchestrate/downgrade/precheck.sls]: true, i will change it

File: salt/metalk8s/orchestrate/downgrade/precheck.sls (hunk 2 above)
Comment (TeddyAndrieux) [reply salt/metalk8s/orchestrate/downgrade/precheck.sls]: > For accuracy's sake we could use `{%- elif salt.pkg.version_cmp(dest_version, values['version']) == 1` here and a relevant failing state name in the `{%- else %}`, where we should theoretically never end up.

Here it's more `{%- elif salt.pkg.version_cmp(dest_version, values['version']) == 0 %}`

> Maybe we can strip labels when we compare versions, in order to actually check `0`, `-1` and `1` explicitly?

I don't get what do you mean ? Btw with `pkg.cmp_version` it's the same
```
[root@single-node ~]# salt-call --local pkg.version_cmp 2.1.1-pw 2.1.1
local:
    0
```
But something weird
```
[root@single-node ~]# salt-call --local pkg.version_cmp 2.1.1-pw 2.1.1-dev
local:
    1
```

But for me `2.1.1` should be superior to `2.1.1-whatever`, as `2.1.1-*` (rc, dev, pw, ... and `2.1.1` should be a released version).

Edit:
But btw looks good as it's alphabetical order `rc` > `pw` > `dev` :stuck_out_tongue:
We just need to handle `*no_suffix*` > `-*`

Comment (TeddyAndrieux) [reply salt/metalk8s/orchestrate/downgrade/precheck.sls]: Done

**Comment by TeddyAndrieux** [comment]
@bert-e reset

**Comment by TeddyAndrieux** [comment]
@bert-e reset

**Review by sayf-eddine-scality — APPROVED** [review]
tested, downgrade successful from 2.2 to 2.1

**Comment by TeddyAndrieux** [comment]
@bert-e approve

**Pull request merged**